File: backend/internal/dao/member_billing.go
package dao

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// MemberBilling links a gym member to a membership plan.
// StripeCustomerID and StripeSubscriptionID are reserved for future Stripe Connect integration.
type MemberBilling struct {
	ID       bson.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	GymID    bson.ObjectID `json:"gym_id" bson:"gym_id"`
	MemberID string        `json:"member_id" bson:"member_id" validate:"required"` // Clerk user ID
	PlanID   bson.ObjectID `json:"plan_id" bson:"plan_id"`

	// Denormalized for display (avoids join on listing)
	PlanName   string `json:"plan_name" bson:"plan_name"`
	MemberName string `json:"member_name" bson:"member_name"`

	Status          string    `json:"status" bson:"status"`           // "active" | "paused" | "cancelled"
	StartDate       time.Time `json:"start_date" bson:"start_date"`
	NextPaymentDate time.Time `json:"next_payment_date" bson:"next_payment_date"`

	// Stripe-ready: populated when Stripe Connect is implemented
	StripeCustomerID     string `json:"stripe_customer_id,omitempty" bson:"stripe_customer_id,omitempty"`
	StripeSubscriptionID string `json:"stripe_subscription_id,omitempty" bson:"stripe_subscription_id,omitempty"`

	CreatedAt time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt time.Time `json:"updated_at" bson:"updated_at"`
}

// PaymentRecord tracks a single billing cycle payment for a member.
// StripePaymentIntentID and StripeInvoiceID are reserved for future Stripe Connect integration.
type PaymentRecord struct {
	ID        bson.ObjectID `json:"id,omitempty" bson:"_id,omitempty"`
	GymID     bson.ObjectID `json:"gym_id" bson:"gym_id"`
	MemberID  string        `json:"member_id" bson:"member_id"`
	BillingID bson.ObjectID `json:"billing_id" bson:"billing_id"`

	// Denormalized for display
	PlanID     bson.ObjectID `json:"plan_id" bson:"plan_id"`
	PlanName   string        `json:"plan_name" bson:"plan_name"`
	MemberName string        `json:"member_name" bson:"member_name"`

	Amount   int64  `json:"amount" bson:"amount"`     // cents
	Currency string `json:"currency" bson:"currency"` // "usd"

	Status  string     `json:"status" bson:"status"`              // "unpaid" | "paid" | "overdue"
	DueDate time.Time  `json:"due_date" bson:"due_date"`
	PaidAt  *time.Time `json:"paid_at,omitempty" bson:"paid_at,omitempty"` // nil until paid

	Notes string `json:"notes,omitempty" bson:"notes,omitempty"`

	// Stripe-ready: populated when Stripe Connect is implemented
	StripePaymentIntentID string `json:"stripe_payment_intent_id,omitempty" bson:"stripe_payment_intent_id,omitempty"`
	StripeInvoiceID       string `json:"stripe_invoice_id,omitempty" bson:"stripe_invoice_id,omitempty"`

	CreatedAt time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt time.Time `json:"updated_at" bson:"updated_at"`
}

File: grapple-frontend/src/api-requests/memberBilling.ts
import { memberBillingApi } from "@/hook/base-apis";

export interface MemberBilling {
  id?: string;
  gym_id: string;
  member_id: string;
  plan_id: string;
  plan_name: string;
  member_name: string;
  status: 'active' | 'paused' | 'cancelled';
  start_date: string;
  next_payment_date: string;
  stripe_customer_id?: string;
  stripe_subscription_id?: string;
  created_at?: string;
  updated_at?: string;
}

export interface PaymentRecord {
  id?: string;
  gym_id: string;
  member_id: string;
  billing_id: string;
  plan_id: string;
  plan_name: string;
  member_name: string;
  amount: number; // cents
  currency: string;
  status: 'unpaid' | 'paid' | 'overdue';
  due_date: string;
  paid_at?: string | null;
  notes?: string;
  stripe_payment_intent_id?: string;
  stripe_invoice_id?: string;
  created_at?: string;
  updated_at?: string;
}

const authHeaders = (token: string) => ({
  headers: { Authorization: `Bearer ${token}` },
});

export const getMemberBilling = async (gymId: string, token: string, memberId?: string): Promise<MemberBilling[]> => {
  const { data } = await memberBillingApi.get<MemberBilling[]>('', {
    params: { gym_id: gymId, ...(memberId ? { member_id: memberId } : {}) },
    ...authHeaders(token),
  });
  return data;
};

export const assignPlan = async (
  payload: { gym_id: string; member_id: string; plan_id: string; member_name: string; start_date?: string },
  token: string
): Promise<{ billing: MemberBilling; payment: PaymentRecord }> => {
  const { data } = await memberBillingApi.post('', payload, authHeaders(token));
  return data;
};

export const updateBillingStatus = async (
  gymId: string,
  billingId: string,
  status: 'active' | 'paused' | 'cancelled',
  token: string
): Promise<MemberBilling> => {
  const { data } = await memberBillingApi.put<MemberBilling>(`/${gymId}/${billingId}`, { status }, authHeaders(token));
  return data;
};

export const getPaymentRecords = async (gymId: string, token: string, memberId?: string, status?: string): Promise<PaymentRecord[]> => {
  const { data } = await memberBillingApi.get<PaymentRecord[]>('/payments', {
    params: {
      gym_id: gymId,
      ...(memberId ? { member_id: memberId } : {}),
      ...(status ? { status } : {}),
    },
    ...authHeaders(token),
  });
  return data;
};

export const updatePaymentStatus = async (
  gymId: string,
  recordId: string,
  payload: { status: 'paid' | 'unpaid' | 'overdue'; notes?: string },
  token: string
): Promise<PaymentRecord> => {
  const { data } = await memberBillingApi.put<PaymentRecord>(`/payments/${gymId}/${recordId}`, payload, authHeaders(token));
  return data;
};
